File: internal/storage/models/game_plays.go
package models

import "time"

// GamePlay represents a single play/action made during a game.
// Actions include card plays, attacks, blocks, land drops, and mulligans.
type GamePlay struct {
	ID             int       `json:"id" db:"id"`
	GameID         int       `json:"game_id" db:"game_id"`
	MatchID        string    `json:"match_id" db:"match_id"`
	TurnNumber     int       `json:"turn_number" db:"turn_number"`
	Phase          string    `json:"phase" db:"phase"`                   // Main1, Combat, Main2, etc.
	Step           string    `json:"step,omitempty" db:"step"`           // BeginCombat, DeclareAttackers, etc.
	PlayerType     string    `json:"player_type" db:"player_type"`       // "player" or "opponent"
	ActionType     string    `json:"action_type" db:"action_type"`       // "play_card", "attack", "block", "land_drop", "mulligan"
	CardID         *int      `json:"card_id,omitempty" db:"card_id"`     // Arena card ID (nullable for some actions)
	CardName       *string   `json:"card_name,omitempty" db:"card_name"` // Card name for display (nullable)
	ZoneFrom       *string   `json:"zone_from,omitempty" db:"zone_from"` // Source zone (hand, library, graveyard, etc.)
	ZoneTo         *string   `json:"zone_to,omitempty" db:"zone_to"`     // Destination zone (battlefield, graveyard, etc.)
	LifeFrom       *int      `json:"life_from,omitempty" db:"life_from"` // Previous life total (for life_change events)
	LifeTo         *int      `json:"life_to,omitempty" db:"life_to"`     // New life total (for life_change events)
	Timestamp      time.Time `json:"timestamp" db:"timestamp"`
	SequenceNumber int       `json:"sequence_number" db:"sequence_number"` // Order within the game
	CreatedAt      time.Time `json:"created_at,omitempty" db:"created_at"`
}

// GameStateSnapshot captures the board state at a specific turn.
type GameStateSnapshot struct {
	ID                  int       `json:"id" db:"id"`
	GameID              int       `json:"game_id" db:"game_id"`
	MatchID             string    `json:"match_id" db:"match_id"`
	TurnNumber          int       `json:"turn_number" db:"turn_number"`
	ActivePlayer        string    `json:"active_player" db:"active_player"` // "player" or "opponent"
	PlayerLife          *int      `json:"player_life,omitempty" db:"player_life"`
	OpponentLife        *int      `json:"opponent_life,omitempty" db:"opponent_life"`
	PlayerCardsInHand   *int      `json:"player_cards_in_hand,omitempty" db:"player_cards_in_hand"`
	OpponentCardsInHand *int      `json:"opponent_cards_in_hand,omitempty" db:"opponent_cards_in_hand"`
	PlayerLandsInPlay   *int      `json:"player_lands_in_play,omitempty" db:"player_lands_in_play"`
	OpponentLandsInPlay *int      `json:"opponent_lands_in_play,omitempty" db:"opponent_lands_in_play"`
	BoardStateJSON      *string   `json:"board_state_json,omitempty" db:"board_state_json"` // JSON snapshot of all permanents on the battlefield
	Timestamp           time.Time `json:"timestamp" db:"timestamp"`
}

// OpponentCardObserved tracks cards revealed by the opponent during a game.
type OpponentCardObserved struct {
	ID            int     `json:"id" db:"id"`
	GameID        int     `json:"game_id" db:"game_id"`
	MatchID       string  `json:"match_id" db:"match_id"`
	CardID        int     `json:"card_id" db:"card_id"`               // Arena card ID
	CardName      *string `json:"card_name,omitempty" db:"card_name"` // Card name for display
	ZoneObserved  string  `json:"zone_observed" db:"zone_observed"`   // Where the card was seen (hand, battlefield, graveyard)
	TurnFirstSeen int     `json:"turn_first_seen" db:"turn_first_seen"`
	TimesSeen     int     `json:"times_seen" db:"times_seen"`
}

// PlayTimelineEntry represents a group of plays during a specific turn/phase.
type PlayTimelineEntry struct {
	Turn     int                `json:"turn"`
	Phase    string             `json:"phase"`
	Plays    []*GamePlay        `json:"plays"`
	Snapshot *GameStateSnapshot `json:"snapshot,omitempty"`
}

// GamePlayFilter provides filtering options for game play queries.
type GamePlayFilter struct {
	MatchID    *string `json:"match_id,omitempty"`
	GameID     *int    `json:"game_id,omitempty"`
	TurnNumber *int    `json:"turn_number,omitempty"`
	PlayerType *string `json:"player_type,omitempty"` // "player" or "opponent"
	ActionType *string `json:"action_type,omitempty"` // "play_card", "attack", "block", "land_drop", "mulligan"
}

// GamePlaySummary provides aggregated play statistics at the match level (across all games).
type GamePlaySummary struct {
	MatchID           string `json:"match_id"`
	TotalPlays        int    `json:"total_plays"`
	PlayerPlays       int    `json:"player_plays"`
	OpponentPlays     int    `json:"opponent_plays"`
	CardPlays         int    `json:"card_plays"`
	Attacks           int    `json:"attacks"`
	Blocks            int    `json:"blocks"`
	LandDrops         int    `json:"land_drops"`
	TotalTurns        int    `json:"total_turns"`
	OpponentCardsSeen int    `json:"opponent_cards_seen"`
}

// Constants for player types.
const (
	PlayerTypePlayer   = "player"
	PlayerTypeOpponent = "opponent"
)

// Constants for action types.
const (
	ActionTypePlayCard   = "play_card"
	ActionTypeAttack     = "attack"
	ActionTypeBlock      = "block"
	ActionTypeLandDrop   = "land_drop"
	ActionTypeMulligan   = "mulligan"
	ActionTypeLifeChange = "life_change"
)

// Constants for game phases.
const (
	PhaseBeginning = "Beginning"
	PhaseMain1     = "Main1"
	PhaseCombat    = "Combat"
	PhaseMain2     = "Main2"
	PhaseEnding    = "Ending"
)

// Constants for combat steps.
const (
	StepBeginCombat      = "BeginCombat"
	StepDeclareAttackers = "DeclareAttackers"
	StepDeclareBlockers  = "DeclareBlockers"
	StepCombatDamage     = "CombatDamage"
	StepEndCombat        = "EndCombat"
)

// Constants for zones.
const (
	ZoneHand        = "hand"
	ZoneLibrary     = "library"
	ZoneBattlefield = "battlefield"
	ZoneGraveyard   = "graveyard"
	ZoneExile       = "exile"
	ZoneStack       = "stack"
	ZoneCommand     = "command"
)

File: frontend/src/services/api/gameplays.ts
/**
 * Game Plays API service.
 *
 * Phase 2 PR #5a: cloud-data play telemetry now hits the BFF directly via
 * apiClient. Routes mount under /api/v1/matches/{matchId}/plays/* and
 * /api/v1/gameplays/game/{gameId}; wire shapes preserved (snake_case keys
 * matching the local TypeScript interfaces below).
 *
 * Plan tracker: .claude/plans/spa-route-migration.md
 */

import { get } from '../apiClient';

/**
 * Represents a single game play action.
 */
export interface GamePlay {
  id: number;
  game_id: number;
  match_id: string;
  turn_number: number;
  phase: string;
  step?: string;
  player_type: 'player' | 'opponent';
  action_type:
    | 'play_card'
    | 'attack'
    | 'block'
    | 'land_drop'
    | 'mulligan'
    | 'life_change'
    | 'cast_spell'
    | 'resolve_spell'
    | 'enter_battlefield'
    | 'to_graveyard'
    | 'exile'
    | 'zone_change';
  card_id?: number;
  card_name?: string;
  zone_from?: string;
  zone_to?: string;
  life_from?: number;
  life_to?: number;
  timestamp: string;
  sequence_number: number;
  created_at?: string;
}

/**
 * Represents a snapshot of game state at a specific turn.
 */
export interface GameStateSnapshot {
  id: number;
  game_id: number;
  match_id: string;
  turn_number: number;
  active_player: string;
  player_life?: number;
  opponent_life?: number;
  player_cards_in_hand?: number;
  opponent_cards_in_hand?: number;
  player_lands_in_play?: number;
  opponent_lands_in_play?: number;
  board_state_json?: string;
  timestamp: string;
}

/**
 * Represents a card observed from the opponent.
 */
export interface OpponentCard {
  id: number;
  game_id: number;
  match_id: string;
  card_id: number;
  card_name?: string;
  zone_observed: string;
  turn_first_seen: number;
  times_seen: number;
}

/**
 * Represents a timeline entry for a turn.
 */
export interface PlayTimelineEntry {
  turn: number;
  active_player: string;
  player_plays: GamePlay[];
  opponent_plays: GamePlay[];
  snapshot?: GameStateSnapshot;
}

/**
 * Represents a summary of plays for a match.
 */
export interface GamePlaySummary {
  match_id: string;
  game_id?: number;
  total_plays: number;
  player_plays: number;
  opponent_plays: number;
  card_plays: number;
  attacks: number;
  blocks: number;
  land_drops: number;
  total_turns: number;
  opponent_cards_seen: number;
}

/**
 * Get all plays for a specific match.
 */
export async function getMatchPlays(matchId: string): Promise<GamePlay[]> {
  return get<GamePlay[]>(`/matches/${encodeURIComponent(matchId)}/plays`);
}

/**
 * Get plays organized by turn for a specific match.
 */
export async function getMatchTimeline(matchId: string): Promise<PlayTimelineEntry[]> {
  return get<PlayTimelineEntry[]>(`/matches/${encodeURIComponent(matchId)}/plays/timeline`);
}

/**
 * Get a summary of plays for a match.
 */
export async function getMatchPlaySummary(matchId: string): Promise<GamePlaySummary> {
  return get<GamePlaySummary>(`/matches/${encodeURIComponent(matchId)}/plays/summary`);
}

/**
 * Get opponent cards observed during a match.
 */
export async function getMatchOpponentCards(matchId: string): Promise<OpponentCard[]> {
  return get<OpponentCard[]>(`/matches/${encodeURIComponent(matchId)}/opponent-cards`);
}

/**
 * Get game state snapshots for a match.
 * Optionally filter by game ID.
 */
export async function getMatchSnapshots(
  matchId: string,
  gameId?: number
): Promise<GameStateSnapshot[]> {
  const url = gameId
    ? `/matches/${encodeURIComponent(matchId)}/snapshots?gameID=${gameId}`
    : `/matches/${encodeURIComponent(matchId)}/snapshots`;
  return get<GameStateSnapshot[]>(url);
}

/**
 * Get plays for a specific game within a match.
 */
export async function getPlaysByGame(gameId: number): Promise<GamePlay[]> {
  return get<GamePlay[]>(`/gameplays/game/${gameId}`);
}

/**
 * Constants for player types.
 */
export const PlayerType = {
  Player: 'player',
  Opponent: 'opponent',
} as const;

/**
 * Constants for action types.
 */
export const ActionType = {
  PlayCard: 'play_card',
  Attack: 'attack',
  Block: 'block',
  LandDrop: 'land_drop',
  Mulligan: 'mulligan',
  LifeChange: 'life_change',
  CastSpell: 'cast_spell',
  ResolveSpell: 'resolve_spell',
  EnterBattlefield: 'enter_battlefield',
  ToGraveyard: 'to_graveyard',
  Exile: 'exile',
  ZoneChange: 'zone_change',
} as const;

/**
 * Constants for game phases.
 */
export const Phase = {
  Beginning: 'Beginning',
  Main1: 'Main1',
  Combat: 'Combat',
  Main2: 'Main2',
  Ending: 'Ending',
} as const;

/**
 * Constants for zone names.
 */
export const Zone = {
  Hand: 'hand',
  Library: 'library',
  Battlefield: 'battlefield',
  Graveyard: 'graveyard',
  Exile: 'exile',
  Stack: 'stack',
  Command: 'command',
} as const;
